pilot/pkg/proxy/envoy/v2/gen2.go

```diff
@@ -0,0 +1,165 @@
+	// a different XDS server instance.
+	nonceSent := w.NonceSent
+
+	// GRPC doesn't send version info in NACKs for RDS. Technically if nonce matches
+	// previous response, it is an ACK/NACK.
+	if nonceSent != "" && nonceSent == discReq.ResponseNonce {
+		adsLog.Debugf("ADS: ACK %s %s %s %v", con.ConID, discReq.VersionInfo, discReq.ResponseNonce,
+			time.Since(w.LastSent))
+		w.NonceAcked = discReq.ResponseNonce
+	}
+
+	if nonceSent != discReq.ResponseNonce {
+		adsLog.Debugf("ADS:RDS: Expired nonce received %s, sent %s, received %s",
+			con.ConID, nonceSent, discReq.ResponseNonce)
+		rdsExpiredNonce.Increment()
+		// This is an ACK for a resource sent on an older stream, or out of sync.
+		// Send a response back.
+		isAck = false
+	}
+
+	// Change in the set of watched resource - regardless of ack, send new data.
+	if !listEqualUnordered(w.ResourceNames, discReq.ResourceNames) {
+		isAck = false
+		w.ResourceNames = discReq.ResourceNames
+	}
+
+	return w, isAck
+}
+
+// handleCustomGenerator uses model.Generator to generate the response.
+func (s *DiscoveryServer) handleCustomGenerator(con *XdsConnection, req *xdsapi.DiscoveryRequest) error {
+	w, isAck := s.handleReqAck(con, req)
+	if isAck {
+		return nil
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     w.TypeUrl,
+		VersionInfo: push.Version, // TODO: we can now generate per-type version !
+		Nonce:       nonce(push.Version),
+	}
+
+	cl := con.node.XdsResourceGenerator.Generate(con.node, push, w)
+	sz := 0
+	for _, rc := range cl {
+		resp.Resources = append(resp.Resources, rc)
+		sz += len(rc.Value)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("ADS: Send failure %s: %v", con.ConID, err)
+		recordSendError(apiSendErrPushes, err)
+		return err
+	}
+	apiPushes.Increment()
+	w.LastSent = time.Now()
+	w.LastSize = sz // just resource size - doesn't include header and types
+	w.NonceSent = resp.Nonce
+
+	return nil
+}
+
+// TODO: verify that ProxyNeedsPush works correctly for Generator - ie. Sidecar visibility
+// is respected for arbitrary resource types.
+
+// Called for config updates.
+// Will not be called if ProxyNeedsPush returns false - ie. if the update
+func (s *DiscoveryServer) pushGeneratorV2(con *XdsConnection, push *model.PushContext, currentVersion string, w *model.WatchedResource) error {
+	// TODO: generators may send incremental changes if both sides agree on the protocol.
+	// This is specific to each generator type.
+	cl := con.node.XdsResourceGenerator.Generate(con.node, push, w)
+	if cl == nil {
+		return nil // No push needed.
+	}
+
+	// TODO: add a 'version' to the result of generator. If set, use it to determine if the result
+	// changed - in many cases it will not change, so we can skip the push. Also the version will
+	// become dependent of the specific resource - for example in case of API it'll be the largest
+	// version of the requested type.
+
+	resp := &xdsapi.DiscoveryResponse{
```

Comment:
There is a lot of duplication here with handleCustomGenerator, does there need to be?

Reply:
Will try to reduce - but it's mostly boilreplate.

Reply:
And far less than the duplication in the handleEDS/LDS/RDS :-)